hazelcast/src/main/java/com/hazelcast/internal/util/JavaVersion.java

```diff
@@ -53,44 +77,83 @@ public static boolean isAtMost(JavaVersion version) {
         return isAtMost(CURRENT_VERSION, version);
     }
 
-    private static JavaVersion detectCurrentVersion() {
-        String version = System.getProperty("java.version");
-        return parseVersion(version);
+    static boolean isAtLeast(JavaMajorVersion currentVersion, JavaMajorVersion minVersion) {
+        return currentVersion.getMajorVersion() >= minVersion.getMajorVersion();
     }
 
-    static JavaVersion parseVersion(String version) {
-        if (version == null) {
-            // this should not happen but it's better to stay on the safe side
-            return UNKNOWN;
-        }
-        JavaVersion result = UNKNOWN;
-        if (version.startsWith("1.")) {
-            String withoutMajor = version.substring(2, version.length());
-            if (withoutMajor.startsWith("6")) {
-                result = JAVA_1_6;
-            } else if (withoutMajor.startsWith("7")) {
-                result = JAVA_1_7;
-            } else if (withoutMajor.startsWith("8")) {
-                result = JAVA_1_8;
+    static boolean isAtMost(JavaMajorVersion currentVersion, JavaMajorVersion maxVersion) {
+        return currentVersion.getMajorVersion() <= maxVersion.getMajorVersion();
+    }
+
+    private static JavaVersion valueOf(int majorVersion) {
+        for (JavaVersion version : values()) {
+            if (version.majorVersion == majorVersion) {
+                return version;
             }
-        } else if (version.startsWith("9")) {
-            // from version 9 the string does not start with "1."
-            result = JAVA_9;
-        } else if (version.startsWith("10")) {
-            result = JAVA_10;
-        } else if (version.startsWith("11")) {
-            result = JAVA_11;
-        } else if (version.startsWith("12")) {
-            result = JAVA_12;
         }
-        return result;
+
+        return null;
     }
 
-    static boolean isAtLeast(JavaVersion currentVersion, JavaVersion minVersion) {
-        return currentVersion.ordinal() >= minVersion.ordinal();
+    private static JavaMajorVersion detectCurrentVersion() {
+        final ILogger logger = Logger.getLogger(JavaVersion.class);
+        final Class runtimeClass;
+        final Class versionClass;
+
+        try {
+            runtimeClass = Class.forName("java.lang.Runtime");
```

Comment:
We don't need reflection here, I think. The `Runtime` is part of Java from its beginning. `Runtime.class` should do the job.

Reply:
Oops. Of course :) Fixed.